keps/sig-storage/2644-honor-pv-reclaim-policy/README.md

```diff
@@ -0,0 +1,380 @@
+PV has a pre-existing `deletionTimestamp`, this makes the method assume that delete is already being processed. This
+results in the plugin volume deletion not being exercised hence leaking the volume.
+```go
+func (ctrl *PersistentVolumeController) deleteVolumeOperation(volume *v1.PersistentVolume) (string, error) {
+	klog.V(4).Infof("deleteVolumeOperation [%s] started", volume.Name)
+
+	// This method may have been waiting for a volume lock for some time.
+	// Previous deleteVolumeOperation might just have saved an updated version, so
+	// read current volume state now.
+	newVolume, err := ctrl.kubeClient.CoreV1().PersistentVolumes().Get(context.TODO(), volume.Name, metav1.GetOptions{})
+	if err != nil {
+		klog.V(3).Infof("error reading persistent volume %q: %v", volume.Name, err)
+		return "", nil
+	}
+
+	if newVolume.GetDeletionTimestamp() != nil {//<==========================DeletionTimestamp set since the PV was deleted first.
+		klog.V(3).Infof("Volume %q is already being deleted", volume.Name)
+		return "", nil
+	}
+    // Remaining code below skipped for brevity...
+    // Please refer pkg/controller/volume/persistentvolume/pv_controller.go in kubernetes/kubernetes for the full code.
+```
+11. Meanwhile, the external-provisioner checks if there is a `deletionTimestamp` on the PV, if so, it assumes that its
+in a transitory state and returns false for the `shouldDelete` check.
+```go
+// shouldDelete returns whether a volume should have its backing volume
+// deleted, i.e. whether a Delete is "desired"
+func (ctrl *ProvisionController) shouldDelete(ctx context.Context, volume *v1.PersistentVolume) bool {
+	if deletionGuard, ok := ctrl.provisioner.(DeletionGuard); ok {
+		if !deletionGuard.ShouldDelete(ctx, volume) {
+			return false
+		}
+	}
+
+	if ctrl.addFinalizer && !ctrl.checkFinalizer(volume, finalizerPV) && volume.ObjectMeta.DeletionTimestamp != nil {
+		return false
+	} else if volume.ObjectMeta.DeletionTimestamp != nil { //<========== DeletionTimestamp is set on the PV since it's deleted first.
+		return false
+	}
+    // Remaining code below skipped for brevity...
+    // Please refer sig-storage-lib-external-provisioner/controller/controller.go for the full code.
+```
+
+The main approach in fixing the issue involves introducing new finalizer `kubernetes.io/pv-deletion-protection` on the Persistent Volume when the controller sees it. Adding the finalizer prevents the PV to be removed from the API server. The finalizer will be removed only after the physical volume on the storage system is deleted.
+
+When the PVC is deleted, the PV is moved into `Released` state and following checks are made:
+
+1. Plugin:
+If the volume has the finalizer `kubernetes.io/pv-deletion-protection`, then, `DeletionTimestamp` checks can be ignored.
+
+3. CSI driver
+
+If when processing the PV update it is observed that it has `kubernetes.io/pv-deletion-protection` finalizer and
```

Comment:
Here I would mention that we're going to use a finalizer already implemented in sig-storage-lib-external-provisioner, which has name `external-provisioner.volume.kubernetes.io/finalizer`.

Reply:
Alternatively, we could add new config option to the library to change the finalizer name.

Reply:
Sure, I'll change this to use the existing finalizer.